File: codecheck/dataset/tester.py
import csv
from csv import DictReader
from enum import Enum
from pathlib import Path
from subprocess import PIPE, Popen, TimeoutExpired
from tempfile import NamedTemporaryFile
from typing import Self

from contexttimer import Timer
from loguru import logger
from pydantic import BaseModel


class Test(BaseModel):
    input: str
    expected: str

    @classmethod
    def read(cls, path: Path) -> list[Self]:
        try:
            text = path.read_text()
        except FileNotFoundError:
            return list()
        reader = DictReader(text.splitlines() or list())
        return [cls.model_validate(row) for row in reader]

    @classmethod
    def write(cls, tests: list[Self], path: Path):
        with (path/"test.csv").open('w') as file:
            writer = csv.writer(file)
            writer.writerow(("input", "expected"))
            writer.writerows(((test.input, test.expected)
                              for test in tests))


class Status(Enum):
    Passed = 'Passed'
    Failed = 'Failed'
    Error = 'Error'


class Result(BaseModel):
    status: Status
    message: str

    @classmethod
    def error(cls, error: str) -> Self:
        return cls(status=Status.Error, message=f"Error:\n{error}")

    @classmethod
    def failed(cls, expected: str, got: str) -> Self:
        return cls(status=Status.Failed, message=f"Fail: {expected=} {got=}")

    @classmethod
    def passed(cls) -> Self:
        return cls(status=Status.Passed, message=f"OK!")


def test(solution: str, tests: list[Test]) -> list[Result]:
    with Timer(output=logger.trace, fmt="Tested in {:.3f} seconds"):
        with NamedTemporaryFile(mode='w') as script:
            print(solution, file=script, flush=True)
            return [
                run(script.name, test)
                for test in tests
            ]


def run(script: str, test: Test) -> Result:
    child = Popen(
        ['python3', script],
        stdin=PIPE, stdout=PIPE, stderr=PIPE,
        text=True,
    )
    try:
        got, error = map(str.strip, child.communicate(test.input, 10))
    except TimeoutExpired:
        child.kill()
        return Result.error("TimeOut")
    if error:
        return Result.error(error)
    if got != test.expected:
        return Result.failed(test.expected, got)
    return Result.passed()

File: WebApp/static/js/tester.js
class Status {
    static Passed = 'Passed';
    static Failed = 'Failed';
    static Error = 'Error';
}

class Result {
    constructor(status, message) {
        this.status = status;
        this.message = message;
    }

    static error(error) {
        return new Result(Status.Error, `Error:\n${error}`);
    }

    static failed(expected, got) {
        return new Result(Status.Failed, `Fail: expected=${expected} got=${got}`);
    }

    static passed() {
        return new Result(Status.Passed, 'OK!');
    }
}


async function run(code, input) {
    console.log(`Gona run ${code} with ${input}`)
    let pyodide = await loadPyodide()
    pyodide.setStdin(new StdinHandler([input]))
    let out = new StdoutHandler()
    pyodide.setStdout({ batched: (msg) => out.add(msg) })
    await pyodide.runPython(code)
    return out.build()
}

async function test(code, input, expected) {
    try {
        let got = await run(code, input)
        if (got === expected)
            return Result.passed()
        else
            return Result.failed(expected, got)
    } catch (error) {
        return Result.error(error)
    }
}

class StdinHandler {
    constructor(lines) {
        this.lines = lines;
        this.idx = 0;
    }

    stdin() {
        return this.lines[this.idx++];
    }
}


class StdoutHandler {
    constructor() {
        this.lines = [];
    }

    add(string) {
        this.lines.push(string)
    }

    build() {
        return this.lines.join('\n');
    }
}